File: M/1911-maximum-alternating-subsequence-sum/main.go
package main

import "fmt"

func maxAlternatingSum(nums []int) int64 {
	s := 0
	var sum int64 = 0

	for s < len(nums) {
		for s < len(nums)-1 {
			if nums[s] <= nums[s+1] {
				s++
			} else {
				break
			}
		}
		sum += int64(nums[s])
		s++

		for s < len(nums)-1 {
			if nums[s] >= nums[s+1] {
				s++
			} else {
				break
			}
		}
		if s < len(nums)-1 {
			sum -= int64(nums[s])
		}
		s++
	}
	return sum
}

func maxAlternatingSumBetter(nums []int) int64 {
	var even int64 = int64(nums[0])
	var odd int64 = 0

	for i := 1; i < len(nums); i++ {
		newodd := odd
		if even-int64(nums[i]) > odd {
			newodd = even - int64(nums[i])
		}
		neweven := even
		if odd+int64(nums[i]) > even {
			neweven = odd + int64(nums[i])
		}
		odd, even = newodd, neweven
	}
	return even
}

func main() {
	fmt.Println()
}

File: M/1911-maximum-alternating-subsequence-sum/main.rs
struct Solution {}

impl Solution {
  pub fn max_alternating_sum(nums: Vec<i32>) -> i64 {
    let mut even: i64 = nums[0] as i64;
    let mut odd: i64 = 0;
    (1..nums.len()).for_each(|n| {
      even = even.max(odd + nums[n] as i64);
      odd = odd.max(even - nums[n] as i64);
    });
    even.max(odd)
  }

  pub fn max_alternating_sum2(nums: Vec<i32>) -> i64 {
    let mut buf: Vec<i64> = Vec::new();
    let mut is_peek: bool = true;
    nums.iter().enumerate().for_each(|(idx, &num)| {
      let mut l: i32 = 0;
      if idx > 0 {
        l = nums[idx - 1];
      }
      let mut r: i32 = 0;
      if idx < nums.len() - 1 {
        r = nums[idx + 1];
      }
      if is_peek && num >= l && num >= r {
        is_peek = false;
        buf.push(num as i64);
      }
      if !is_peek && num <= l && num <= r {
        is_peek = true;
        buf.push((-num) as i64);
      }
    });
    // println!("{:?}",buf);
    if buf.len() % 2 == 0 {
      buf.pop();
    }
    buf.iter().sum::<i64>()
  }
}
